### ISSUE TITLE
Snap points got all messed up

### ISSUE THREAD

**Issue by ndavalos** [EpochModTeam/DayZ-Epoch#1812]
I've had a lot of complaints about snap points, snapping is slightly off, the orbs for the snap points are also way off too when displayed.

http://images.akamai.steamusercontent.com/ugc/99473989837893280/7233856DF9EF919EBBB9FED76FB020573DEFCE17/

http://images.akamai.steamusercontent.com/ugc/99473989837897719/DE3715F7E47A56ECE6FE7AAAA521F4DDB775FF05/

http://images.akamai.steamusercontent.com/ugc/99473989837897492/A4B4D52A78ED732F11151A527FA8AE22628789EE/

http://images.akamai.steamusercontent.com/ugc/99473989837897943/33F3BD63030E5CCE56D2CA79D28C7086AA2B4673/

These were perfect for the most part a while back, not sure what happened.

**Comment by oiad** [comment]
Is that just metal floors alone? Or is there any other build able nearby?

**Comment by ndavalos** [comment]
There were cinder walls under that. I haven't gotten a good answer from my players about what is messed up, I've asked for a list of things. I haven't had time to take a look for myself. I'' trying to figure out why the fps has dropped so dramatically in the last month and a half too. It's abysmal.

**Comment by oiad** [comment]
I tested all the snap stuff quite a lot, already have some big bases on my server from players, no-one has reported anything (Not to say there isnt a problem) I'll try building some metal floors tonight and see where I end up.

I find the server fps fluctuates wildly sometimes, what are you hardware wise?

**Comment by ndavalos** [comment]
Server is fine, it's all client side. Server is i7 6700k oc with ssds. My client is an i7 4790k, I was using compression on the mission with makepbo, got like 35fps. Turned off compression and it's fluctuation between 40 and 50. Most players are getting 20 and below with massive frame drops and crashes. The crashes and frame drops may be battleye related, which I've seen before. 105 I had 60fps all the time for the most part. October I was also getting 60fps with 106. A lot of stuff got changed around in the last month and a half and the frame rate kept dropping. I have to sit down and investigate further, but I did notice that turning off the key bind check got a slight boost. I'm wondering what else might be lurking.

**Comment by ebayShopper** [comment]
In regards to FPS, I recently noticed the points of interest buildings on Chernarus are currently always present and global spawned. It spawns a massive number of map objects (clutter cutter, plants, buildings, etc.).

If you set `dayz_POIs = false;` in init.sqf you should see a big FPS boost. I will test changing these back to local spawned and probably disable them by default.

**Comment by ndavalos** [comment]
I turned off poi on chernarus long time ago. I got semi decent fps there after disabling all compression with makepbo. Everything else though, namalsk, especially taviana are just horrendous for most players.

**Comment by ndavalos** [comment]
As a side note, I also spawn almost everything server side including traders. Really the only thing that happens client side is zombies.

**Comment by ndavalos** [comment]
My admin just mentioned he's been hopping around to other servers and most of them are worse than mine in terms of fps client side. I'' considering removing most of the dialog/control battleye filters to see if that helps some.

**Comment by ebayShopper** [comment]
If you are referring to the DZE_FilterCheats check it only runs on the default display when NumPad-, PushToTalk or VON keys are pressed, so it shouldn't affect client FPS during normal play.

I have also noticed more client crashes recently than in the past. It happens to me mainly when tabbing out of the game and back in again. That could be due to a recent BE update though if it is happening on vanilla and 1051 servers too. Scripts.txt was supposedly optimized recently according to BE, but it was known to cause client FPS drops in the past.

Not sure why FPS would be worse on alternative maps. If anything it should be better, since there are fewer map objects being spawned.

**Comment by ndavalos** [comment]
It was the checks for key binds for turbo, hold breath etc. I just commented all that out and left the voice stuff in. ui_update[whatever] the file is. I don't know if it was coincidence or not honestly. I have a little reservation about the oneachframe thing happening. I think you're updating group icons there, I forget the file name. I took the group icon updates out of there since I have my own group system, but there's other things like that I'm going to have to examine more closely, like the client scheduler too.

**Comment by ndavalos** [comment]
Ok, so about the snap, my admin is saying anything build pre 106 doesn't have snap points working properly. I'll have to check this out, not sure why except exact building might be interfering with it and vectors?

**Comment by oiad** [comment]
It gets the snap points based off the class name so pre 106 or 106 _should_ not matter, maybe.

I have noticed that the snap points gets confused based off objects around it, I.e if you're building inside a room it can snap to the wrong object. So you have to manually set the snap point.

Those snap points are really hard to get right especially since you have to reload the server each time to test, would be nice if someone else had a nosy at them, I've tried my best

**Comment by ebayShopper** [comment]
Ui_updateControls only runs once at login and once each time you close the Controls dialog in your options menu. It isn't a loop, so it has no affect on FPS.

The group icons do loop onEachFrame if you have dayz_groupSystem=true. Unfortunately there is no way to draw group icons on each player in A2 without a loop, since Arma's built in group icons only draw on the leader. Anything less than a 0.001s (same as onEachFrame) loop is noticeably delayed updating the position. I've made the loop as efficient as I know how, but pull requests or suggestions for optimization are always welcome.

**Comment by ndavalos** [comment]
The original snap building worked much better, this was not an issue prior to 106. Having things off at slight angles shouldn't be happening.

It's pretty much just like what's shown in the screenshots for anything pre 106 being snapped to from what I'm being told unfortunately.

**Comment by ndavalos** [comment]
oneachframe is a really harsh way to do anything really. I wish I knew where things went wrong, I might have to pull a copy of the repository from a point in time that I think it was running well and start from there. I also have to make sure all my stuff merged properly, like the zombie spawn thing, I saw you moved it to the init.

My players are getting errors about missing properties on the graves_light skin before crashes. I'm not sure if it specific to that skin or not. Waiting on an rpt from someone.

**Comment by icomrade** [comment]
You should be using Graves_Light_DZ, I assume you are, the only change was a fix for SetFace errors. this was the only line added https://github.com/EpochModTeam/DayZ-Epoch/blob/ea389b59eaa8074816e7f5b121d3eeada4f9644b/SQF/dayz_code/Configs/CfgVehicles/Skins.hpp#L734

**Comment by ndavalos** [comment]
@icomrade yeah I'm not sure what's up with that. I'm thinking it's a battleye memory leak and that's actually a side effect. Hopefully I'll have more information later today or tomorrow.

**Comment by SmokeyBR** [comment]
looking at the pictures looks like he was turning with Q or E with a degree set when `DZE_dirWithDegrees = true;`
although it does not happen right after u snap to an object

**Comment by ndavalos** [comment]
I was, but it was with default degree settings, and it should have rotated properly to the right position as it did in 105.

**Comment by icomrade** [comment]
Any change you would know your server bandwidth usage at times of poor fps? I think we may greatly benefit from leaving simulation disabled  on vehicles until a player is near, checking probably every 1-2 seconds on the server.

**Comment by ndavalos** [comment]
Bandwidth has been fine and really low. 106 killed my population, so I haven't had enough people on to see any high bandwidth usage.

**Comment by oiad** [comment]
As with what @SmokeyBR  said, I haven't touched any of the rotation angles only the snappoints which shouldn't affect rotation

**Comment by ndavalos** [comment]
It depends, I'll have to look at the code later, but I do know the original vector and snap building mod in 105 worked perfectly.

**Comment by ndavalos** [comment]
Ok, got an update on snapping from my admin, it works fine until a restart, after that you get what's in the screenshots for results if you try to build more.

**Comment by ndavalos** [comment]
Something else interesting from my admin's rpt, I have to see if snap does this in mine, but he's getting a lot of Ref to nonnetwork object 42ee5d00# 738126: sphere10cm_ep1.p3d on tavi. I'm pretty sure those are the little orbs for snap points.

**Comment by oiad** [comment]
They are annoying but normal. It's due to them being deleted when you finish building

**Comment by icomrade** [comment]
Ref to nonnetwork object means that a global command was executed on a non-global object. in this case, since there is no deletevehiclelocal, deletevehicle networks the delete request to all players and the server, and it throws that message since it's local to another player.

**Comment by looter809** [comment]
Could the FPS drop be from the dayz_clientpreload error spam? `Error in expression <dayz_clientPreload and !(isNull _agent)>
  Error position: <dayz_clientPreload and !(isNull _agent)>
  Error Undefined variable in expression: dayz_clientpreload`
In my AppData\Local\ArmA 2 OA\ArmA2OA.RPT I am getting this message spammed nonstop. Is this the preload that DayZ Launcher uses? I have thousands of spam messages from that clientpreload. My players are reporting about half the FPS they use to get. I was personally getting 70 FPS on my test server before the update and with nothing on the test server (base/player data wise) and now I am getting 30-35 FPS.

Maybe this helps, maybe it doesn't. I noticed the error was fixed in the current test build, but I have not tried the test build since 1.0.6a.
If someone finds out how to fix the FPS, I hope I hear about it xD

Thanks Epoch team for all the hard work!

**Comment by looter809** [comment]
Another thing that may want to be changed is the price of default DMR and M24 mags. `class 5Rnd_762x51_M24 {
		type = "trade_items";
		buy[] = {1,"ItemSilverBar10oz"};
		sell[] = {5,"ItemSilverBar"};
	};
	class 20Rnd_762x51_DMR {
		type = "trade_items";
		buy[] = {6,"ItemGoldBar"};
		sell[] = {3,"ItemGoldBar"};`
Find a player duping money by this method earlier by combining m24 mags.

**Comment by ndavalos** [comment]
So here's a video from my admin demonstrating:
https://www.youtube.com/watch?v=60wtAS8HhgY&feature=youtu.be

The first floor he places i on a wall after a restart. The rest of the video shows him using snap on new building objects. As you can see the objects that existed before a restart snap weird. Newly placed objects snap correctly.

This one is just newly placed objects:
https://www.youtube.com/watch?v=Avw-VXEBfXE&feature=youtu.be

**Comment by ebayShopper** [comment]
@looter809 That error only happens on the main menu screen during intro.sqs because it is spawning zombies before the variable is initialized. It will stop as soon as you load into a server or use -world=empty launch parameter so intro.sqs doesn't run. Already fixed for 1.0.6.1 with 2db6692.

Fixed the M24 and DMR ammo price with 0d4d288, thanks. For any other problems please create a separate issue, since this one is getting off topic.

**Comment by icomrade** [comment]
I can't reproduce the snap issue on my test server, which is on Chernarus. It looks like the degrees is set to something <= 5 degrees  and the player hit a key that would rotate the object. I'll continue trying to replicate, but so far no issues.

**Comment by ndavalos** [comment]
@icomrade, remember the snapping doesn't seem to break until after a restart. Which would explain it being slightly off with a converted database. I'LOL try to figure it out too this weekend.

**Comment by icomrade** [comment]
Yup tried before and after a restart

**Comment by Abe-Andet** [comment]
Any more regarding the snap?  Reports over several servers.  The snap feature works fine when starting a new base.  Everything snaps like it used to in 105.   However.. if you snap a new object to a pre-existing object it will snap to it.. but at a weird angle instead of parallel or 90 degree.  The new object seems to snap to an old object with a North/south orientation.  There is no set "off degree" because of the North/south orientation.  This includes after a restart.   Any object that is there after a restart is now considered a pre-existing object.  So, trying to snap to that "old " object results in the new object being off again.

**Comment by oiad** [comment]
I can't reproduce this on my server, I'm using chernarus and default everything epoch related.

Do you have any steps to reproduce @Abe-Andet apart from what has been said? What map? Converted database? Any location idea that you were so we can try and see if it's a terrain issue?

**Comment by Abe-Andet** [comment]
Sorry haven't gotten back.  It isn't a terrain issue and not map specific.   The same thing that happens in the videos above...    https://www.youtube.com/watch?v=60wtAS8HhgY&feature=youtu.be  

The snapping works fine until  a restart or when placing on old items.   The new items snap to an old item fine...  just not at a parallel or 90 degree.   In 105 and previous any parts would snap at parallel or 90 and you just hit the Q or E twice (45 degree default) or  once if set to 90   The parts are always facing to the cardinal points on a map when put on an old object.

@oiad can i try it on your server?

**Comment by ndavalos** [comment]
You can close this, it was my fault for not broadcasting memDir in server_monitor.sqf
